File: 分省数据看板（月度）/趋势分析/build_trend_data.py
# -*- coding: utf-8 -*-
"""
从 趋势分析 文件夹下的各月 Excel 读取全国省份核心指标，生成 trend-data.js 供看板使用。
运行：在「分省数据看板（月度）」目录下执行  python 趋势分析/build_trend_data.py
输出：趋势分析/trend-data.js
"""
import os
import re
import json
import glob

try:
    import pandas as pd
except ImportError:
    print('请先安装 pandas 和 openpyxl: pip install pandas openpyxl')
    exit(1)

# 当前脚本所在目录 = 趋势分析/
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
OUTPUT_JS = os.path.join(SCRIPT_DIR, 'trend-data.js')

# 可能的列名映射（Excel 列名 -> 标准字段）
PROVINCE_KEYS = ['地区', '省份', '省', '区域', 'province']
ACTIVE_KEYS = ['活跃用户', '月活', '活跃']
REVENUE_KEYS = ['订单营收', '营收', '收入', 'revenue']
RATE_KEYS = ['使用率', '使用率%', '渗透率']
ARPU_KEYS = ['ARPU', 'arpu']


def find_column(row, keys):
    """在表头行中查找匹配的列名（不区分大小写）。"""
    for c in row:
        if c is None or (isinstance(c, float) and pd.isna(c)):
            continue
        s = str(c).strip()
        for k in keys:
            if k in s or s in k:
                return c
    return None


def read_month_sheet(path):
    """读取单月 Excel，返回 [(省份, 活跃用户, 营收, 使用率, ARPU), ...]"""
    df = pd.read_excel(path, sheet_name=0, header=None)
    if df.empty or len(df) < 2:
        return []
    # 第一行作为表头
    header = df.iloc[0]
    col_prov = find_column(header, PROVINCE_KEYS)
    col_active = find_column(header, ACTIVE_KEYS)
    col_rev = find_column(header, REVENUE_KEYS)
    col_rate = find_column(header, RATE_KEYS)
    col_arpu = find_column(header, ARPU_KEYS)
    if col_prov is None:
        # 尝试第一列作为省份
        col_prov = header.iloc[0] if len(header) > 0 else None
    prov_idx = list(header).index(col_prov) if col_prov is not None else 0
    active_idx = list(header).index(col_active) if col_active is not None else None
    rev_idx = list(header).index(col_rev) if col_rev is not None else None
    rate_idx = list(header).index(col_rate) if col_rate is not None else None
    arpu_idx = list(header).index(col_arpu) if col_arpu is not None else None

    rows = []
    for i in range(1, len(df)):
        row = df.iloc[i]
        prov = row.iloc[prov_idx] if prov_idx is not None else None
        if prov is None or (isinstance(prov, float) and pd.isna(prov)):
            continue
        prov = str(prov).strip()
        if not prov or prov.startswith('同比') or prov.startswith('环比'):
            continue
        try:
            active = float(row.iloc[active_idx]) if active_idx is not None and active_idx < len(row) else 0
        except (ValueError, TypeError):
            active = 0
        try:
            rev = float(row.iloc[rev_idx]) if rev_idx is not None and rev_idx < len(row) else 0
        except (ValueError, TypeError):
            rev = 0
        try:
            rate = float(row.iloc[rate_idx]) if rate_idx is not None and rate_idx < len(row) else 0
        except (ValueError, TypeError):
            rate = 0
        try:
            arpu = float(row.iloc[arpu_idx]) if arpu_idx is not None and arpu_idx < len(row) else 0
        except (ValueError, TypeError):
            arpu = 0
        rows.append((prov, active, rev, rate, arpu))
    return rows


def main():
    # 匹配 25年1月.xlsx ~ 26年2月.xlsx
    pattern = os.path.join(SCRIPT_DIR, '*.xlsx')
    files = sorted(glob.glob(pattern))
    month_files = []
    for f in files:
        base = os.path.basename(f)
        m = re.match(r'(\d{2})年(\d{1,2})月\.xlsx', base)
        if m:
            month_files.append((f, f'{m.group(1)}年{m.group(2)}月'))

    if not month_files:
        print('未找到 趋势分析/*.xlsx 文件')
        return

    # 按月份排序（25年1月 ... 26年2月）
    def month_key(t):
        a, b = re.match(r'(\d{2})年(\d{1,2})月', t[1]).groups()
        return (int(a), int(b))
    month_files.sort(key=month_key)

    # 构建 provinceTrendData: { 省份: [ {月份, 活跃用户, 营收, 使用率, ARPU}, ... ] }
    province_trend = {}
    # 构建 coreData: { 月份: { 省份: { 活跃用户, 订单营收, 使用率, ARPU, ... } } }
    core_data = {}

    for path, month_label in month_files:
        core_data[month_label] = {}
        rows = read_month_sheet(path)
        for prov, active, rev, rate, arpu in rows:
            if prov not in province_trend:
                province_trend[prov] = []
            province_trend[prov].append({
                '月份': month_label,
                '活跃用户': int(active) if active == int(active) else round(active, 2),
                '营收': round(rev, 2) if isinstance(rev, float) else rev,
                '使用率': round(rate, 1) if isinstance(rate, float) else rate,
                'ARPU': round(arpu, 2) if isinstance(arpu, float) else arpu
            })
            core_data[month_label][prov] = {
                '活跃用户': int(active) if active == int(active) else round(active, 2),
                '订单营收': round(rev, 2) if isinstance(rev, float) else rev,
                '使用率': round(rate, 1) if isinstance(rate, float) else rate,
                'ARPU': round(arpu, 2) if isinstance(arpu, float) else arpu,
                '深度访问率': 0, '新用户': 0, '老用户': 0,
                '同比活跃': 0, '环比活跃': 0, '同比营收': 0, '环比营收': 0,
                '同比使用率': 0, '环比使用率': 0, '同比ARPU': 0, '环比ARPU': 0
            }

    # 保证每个省份的趋势按月份顺序一致
    all_months = [m for _, m in month_files]
    for prov in province_trend:
        by_month = {x['月份']: x for x in province_trend[prov]}
        province_trend[prov] = []
        for m in all_months:
            if m in by_month:
                province_trend[prov].append(by_month[m])
            else:
                province_trend[prov].append({
                    '月份': m, '活跃用户': 0, '营收': 0, '使用率': 0, 'ARPU': 0
                })

    provinces = sorted(province_trend.keys())
    print(f'共 {len(provinces)} 个省份/地区，{len(all_months)} 个月份')
    print('省份:', provinces[:10], '...' if len(provinces) > 10 else '')

    # 输出 trend-data.js
    js_content = '''// 由 趋势分析/build_trend_data.py 自动生成，请勿直接编辑
// 全国各省份趋势数据与核心指标（25年1月～26年2月）

const TREND_PROVINCES = ''' + json.dumps(provinces, ensure_ascii=False, indent=2) + ''';

const provinceTrendData = ''' + json.dumps(province_trend, ensure_ascii=False, indent=2) + ''';

const coreDataFromTrend = ''' + json.dumps(core_data, ensure_ascii=False, indent=2) + ''';

const TREND_MONTHS = ''' + json.dumps(all_months, ensure_ascii=False) + ''';
'''

    with open(OUTPUT_JS, 'w', encoding='utf-8') as f:
        f.write(js_content)
    print('已生成:', OUTPUT_JS)


if __name__ == '__main__':
    main()

File: 分省数据看板（月度）/趋势分析/build_trend_data.js
/**
 * 从 趋势分析 文件夹下的各月 Excel 读取全国省份核心指标，生成 trend-data.js 供看板使用。
 * 运行：在「分省数据看板（月度）」目录下执行  node 趋势分析/build_trend_data.js
 * 或  npm run build-trend
 */
const fs = require('fs');
const path = require('path');

let XLSX;
try {
  XLSX = require('xlsx');
} catch (e) {
  console.error('请先安装依赖: npm install');
  process.exit(1);
}

const SCRIPT_DIR = __dirname;
// 输出到看板根目录，与 index.html 同目录，避免 file:// 下路径问题
const OUTPUT_JS = path.join(SCRIPT_DIR, '..', 'trend-data.js');

// 表头对应：省份名称、活跃用户、新用户、营收、使用用户、ARPU、使用率
const PROVINCE_KEYS = ['省份名称', '地区', '省份', '省', '区域', 'province'];
const ACTIVE_KEYS = ['活跃用户'];
const NEW_USER_KEYS = ['新用户'];
const REVENUE_KEYS = ['营收', '订单营收', '收入', 'revenue'];
const RATE_KEYS = ['使用率', '使用率%', '渗透率'];
const ARPU_KEYS = ['ARPU', 'arpu'];

/** 统一省份名称，避免「山东」与「山东省」同时出现 */
function normalizeProvinceName(name) {
  if (!name || typeof name !== 'string') return name;
  const s = name.trim();
  if (s.endsWith('省')) return s.slice(0, -1);
  if (s === '广西壮族自治区' || s === '广西') return '广西';
  if (s === '内蒙古自治区' || s === '内蒙古') return '内蒙古';
  if (s === '新疆维吾尔自治区' || s === '新疆') return '新疆';
  if (s === '宁夏回族自治区' || s === '宁夏') return '宁夏';
  if (s === '西藏自治区' || s === '西藏') return '西藏';
  return s;
}

function findColumnIndex(header, keys) {
  for (let i = 0; i < header.length; i++) {
    const cell = header[i];
    if (cell == null || (typeof cell === 'number' && isNaN(cell))) continue;
    const s = String(cell).trim();
    for (const k of keys) {
      if (s.includes(k) || k.includes(s)) return i;
    }
  }
  return -1;
}

/** 解析单元格数值：支持纯数字、带%的字符串、千分位逗号、Excel 小数率(0.72→72) */
function parseNumber(val, opts = {}) {
  if (val == null || val === '') return 0;
  const { asRate } = opts; // 使用率等，Excel 可能存为 0.72 表示 72%
  if (typeof val === 'number') {
    if (!isNaN(val) && asRate && val > 0 && val <= 1) return Math.round(val * 1000) / 10;
    return isNaN(val) ? 0 : val;
  }
  const s = String(val).trim().replace(/,/g, '');
  if (s === '') return 0;
  if (s.endsWith('%')) {
    const n = Number(s.slice(0, -1));
    return isNaN(n) ? 0 : n;
  }
  const n = Number(s);
  if (!isNaN(n) && asRate && n > 0 && n <= 1) return Math.round(n * 1000) / 10;
  return isNaN(n) ? 0 : n;
}

function parseMonthLabel(label) {
  const m = String(label).match(/^(\d{2})年(\d{1,2})月$/);
  if (!m) return null;
  return { year: parseInt(m[1], 10), month: parseInt(m[2], 10) };
}

function formatMonthLabel(year, month) {
  const yy = String(year).padStart(2, '0').slice(-2);
  return `${yy}年${month}月`;
}

function prevMonthLabel(label) {
  const p = parseMonthLabel(label);
  if (!p) return null;
  if (p.month === 1) return formatMonthLabel(p.year - 1, 12);
  return formatMonthLabel(p.year, p.month - 1);
}

function yoyMonthLabel(label) {
  const p = parseMonthLabel(label);
  if (!p) return null;
  return formatMonthLabel(p.year - 1, p.month);
}

function readMonthSheet(filePath) {
  const wb = XLSX.readFile(filePath, { cellDates: false, raw: false });
  const sheetName = wb.SheetNames[0];
  const ws = wb.Sheets[sheetName];
  const data = XLSX.utils.sheet_to_json(ws, { header: 1, defval: null });
  if (!data || data.length < 2) return [];

  const header = data[0];
  let provIdx = findColumnIndex(header, PROVINCE_KEYS);
  if (provIdx < 0) provIdx = 0;
  const activeIdx = findColumnIndex(header, ACTIVE_KEYS);
  const newUserIdx = findColumnIndex(header, NEW_USER_KEYS);
  const revIdx = findColumnIndex(header, REVENUE_KEYS);
  const rateIdx = findColumnIndex(header, RATE_KEYS);
  const arpuIdx = findColumnIndex(header, ARPU_KEYS);

  const rows = [];
  for (let i = 1; i < data.length; i++) {
    const row = data[i];
    let prov = row[provIdx];
    if (prov == null || (typeof prov === 'number' && isNaN(prov))) continue;
    prov = normalizeProvinceName(String(prov).trim());
    if (!prov || prov.startsWith('同比') || prov.startsWith('环比')) continue;

    const num = (idx, opts) => {
      if (idx < 0 || idx >= row.length) return 0;
      return parseNumber(row[idx], opts);
    };
    const numOptional = (idx) => {
      if (idx < 0 || idx >= row.length) return null;
      const val = row[idx];
      if (val == null || String(val).trim() === '') return null;
      return parseNumber(val);
    };

    rows.push({
      prov,
      active: num(activeIdx),
      newUser: numOptional(newUserIdx),
      rev: num(revIdx),
      rate: num(rateIdx, { asRate: true }),
      arpu: num(arpuIdx),
    });
  }
  return rows;
}

function main() {
  const inspect = process.argv.includes('--inspect');
  const files = fs.readdirSync(SCRIPT_DIR)
    .filter((f) => /^\d{2}年\d{1,2}月\.xlsx$/.test(f))
    .sort((a, b) => {
      const ma = a.match(/^(\d{2})年(\d{1,2})月/);
      const mb = b.match(/^(\d{2})年(\d{1,2})月/);
      if (!ma || !mb) return 0;
      const na = parseInt(ma[1], 10) * 100 + parseInt(ma[2], 10);
      const nb = parseInt(mb[1], 10) * 100 + parseInt(mb[2], 10);
      return na - nb;
    });

  if (files.length === 0) {
    console.log('未找到 趋势分析/*.xlsx 文件');
    return;
  }

  if (inspect) {
    const firstPath = path.join(SCRIPT_DIR, files[0]);
    const wb = XLSX.readFile(firstPath);
    const data = XLSX.utils.sheet_to_json(wb.Sheets[wb.SheetNames[0]], { header: 1, defval: null });
    const header = data[0] || [];
    console.log('【调试】文件:', files[0]);
    console.log('表头(第1行):', header);
    console.log('前3行数据:', data.slice(1, 4));
    console.log('识别到的列索引 - 省份:', findColumnIndex(header, PROVINCE_KEYS), '活跃:', findColumnIndex(header, ACTIVE_KEYS), '新用户:', findColumnIndex(header, NEW_USER_KEYS), '营收:', findColumnIndex(header, REVENUE_KEYS), '使用率:', findColumnIndex(header, RATE_KEYS), 'ARPU:', findColumnIndex(header, ARPU_KEYS));
    return;
  }

  const monthFiles = files.map((f) => {
    const m = f.match(/^(\d{2})年(\d{1,2})月/);
    return [path.join(SCRIPT_DIR, f), m ? `${m[1]}年${m[2]}月` : ''];
  }).filter(([, label]) => label);

  const provinceTrend = {};
  const coreData = {};

  for (const [filePath, monthLabel] of monthFiles) {
    coreData[monthLabel] = {};
    const rows = readMonthSheet(filePath);
    for (const { prov, active, newUser, rev, rate, arpu } of rows) {
      if (!provinceTrend[prov]) provinceTrend[prov] = [];
      const fmtInt = (v) => (v == null ? null : (Number.isInteger(v) ? v : Math.round(v * 100) / 100));
      provinceTrend[prov].push({
        月份: monthLabel,
        活跃用户: fmtInt(active) ?? 0,
        新用户: fmtInt(newUser),
        营收: Math.round(rev * 100) / 100,
        使用率: Math.round(rate * 10) / 10,
        ARPU: Math.round(arpu * 100) / 100,
      });
      coreData[monthLabel][prov] = {
        活跃用户: fmtInt(active) ?? 0,
        新用户: fmtInt(newUser),
        订单营收: Math.round(rev * 100) / 100,
        使用率: Math.round(rate * 10) / 10,
        ARPU: Math.round(arpu * 100) / 100,
        深度访问率: 0,
        老用户: 0,
        同比活跃: null,
        环比活跃: null,
        同比新用户: null,
        环比新用户: null,
        同比营收: null,
        环比营收: null,
        同比使用率: null,
        环比使用率: null,
        同比ARPU: null,
        环比ARPU: null,
      };
    }
  }

  const allMonths = monthFiles.map(([, m]) => m);
  for (const prov of Object.keys(provinceTrend)) {
    const byMonth = {};
    provinceTrend[prov].forEach((item) => { byMonth[item.月份] = item; });
    provinceTrend[prov] = allMonths.map((m) =>
      byMonth[m] || { 月份: m, 活跃用户: 0, 新用户: null, 营收: 0, 使用率: 0, ARPU: 0 }
    );
  }

  // 根据趋势序列计算同比、环比（按月份标签匹配，非数组下标）
  const pctChange = (cur, prev) => {
    if (prev == null || prev === 0) return null;
    return Math.round((cur - prev) / prev * 100);
  };
  const ppChange = (cur, prev) => {
    if (prev == null) return null;
    return Math.round((cur - prev) * 10) / 10;
  };
  for (const prov of Object.keys(provinceTrend)) {
    const arr = provinceTrend[prov];
    const byMonth = {};
    arr.forEach((item) => { byMonth[item.月份] = item; });
    for (let i = 0; i < arr.length; i++) {
      const m = arr[i].月份;
      if (!coreData[m] || !coreData[m][prov]) continue;
      const cur = arr[i];
      const prevMonth = byMonth[prevMonthLabel(m)] || null;
      const lastYear = byMonth[yoyMonthLabel(m)] || null;
      coreData[m][prov].同比活跃 = lastYear ? pctChange(cur.活跃用户, lastYear.活跃用户) : null;
      coreData[m][prov].环比活跃 = prevMonth ? pctChange(cur.活跃用户, prevMonth.活跃用户) : null;
      coreData[m][prov].同比新用户 = (lastYear && cur.新用户 != null && lastYear.新用户 != null)
        ? pctChange(cur.新用户, lastYear.新用户) : null;
      coreData[m][prov].环比新用户 = (prevMonth && cur.新用户 != null && prevMonth.新用户 != null)
        ? pctChange(cur.新用户, prevMonth.新用户) : null;
      coreData[m][prov].同比营收 = lastYear ? pctChange(cur.营收, lastYear.营收) : null;
      coreData[m][prov].环比营收 = prevMonth ? pctChange(cur.营收, prevMonth.营收) : null;
      coreData[m][prov].同比使用率 = lastYear ? ppChange(cur.使用率, lastYear.使用率) : null;
      coreData[m][prov].环比使用率 = prevMonth ? ppChange(cur.使用率, prevMonth.使用率) : null;
      coreData[m][prov].同比ARPU = lastYear ? pctChange(cur.ARPU, lastYear.ARPU) : null;
      coreData[m][prov].环比ARPU = prevMonth ? ppChange(cur.ARPU, prevMonth.ARPU) : null;
    }
  }

  const provinces = Object.keys(provinceTrend).sort();
  console.log('共', provinces.length, '个省份/地区，', allMonths.length, '个月份');
  console.log('省份:', provinces.slice(0, 10).join('、'), provinces.length > 10 ? '...' : '');

  const jsContent = `// 由 趋势分析/build_trend_data.js 自动生成，请勿直接编辑
// 全国各省份趋势数据与核心指标（25年1月～26年2月）
// 使用 window 挂载，保证在 file:// 下被 app.js 正确读取

window.TREND_PROVINCES = ${JSON.stringify(provinces, null, 2)};
window.provinceTrendData = ${JSON.stringify(provinceTrend, null, 2)};
window.coreDataFromTrend = ${JSON.stringify(coreData, null, 2)};
window.TREND_MONTHS = ${JSON.stringify(allMonths, null, 2)};
`;

  fs.writeFileSync(OUTPUT_JS, jsContent, 'utf8');
  console.log('已生成:', OUTPUT_JS);
}

main();
